test/test_logManager.py

```diff
@@ -0,0 +1,1155 @@
+    # should begin calculation and ensure that the true time is
+    # within a tolerance of the estimated time
+    tol = 0.3
+
+    test_timer = ETA(50, "t_fin")
+    basicLogmgr.add_quantity(test_timer)
+
+    sleepTime = 0.1
+
+    # add first tick
+    basicLogmgr.tick_before()
+    sleep(sleepTime)
+    basicLogmgr.tick_after()
+
+    # add second tick
+    basicLogmgr.tick_before()
+    sleep(sleepTime)
+    basicLogmgr.tick_after()
+
+    N = 30
+    last = basicLogmgr.get_expr_dataset("t_fin")[-1][-1][-1]
+
+    for i in range(N):
+        basicLogmgr.tick_before()
+        sleep(sleepTime)
+        basicLogmgr.tick_after()
+
+        actual_time = basicLogmgr.get_expr_dataset("t_fin")[-1][-1][-1]
+        print(last, actual_time)
+        # assert that these quantities only
+        # differ by a max of tol defined above
+        if i > 5:  # dont expect the first couple to be accurate
+            assert abs(last - actual_time) < tol
+        last = last - sleepTime
+
+
+def test_MemoryHwm_quantity(basicLogmgr: LogManager):
+    # can only check if nothing breaks and the watermark never lowers,
+    # as we do not know what else is on the system
+
+    # set a run property
+    basicLogmgr.set_constant("myconst", uniform(0, 1))
+
+    # Generic run metadata, such as command line, host, and time
+    add_run_info(basicLogmgr)
+
+    # Time step duration, wall time, ...
+    add_general_quantities(basicLogmgr)
+
+    # Simulation time, time step
+    add_simulation_quantities(basicLogmgr)
+
+    # Additional quantities to log
+    vis_timer = IntervalTimer("t_vis", "Time spent visualizing")
+    basicLogmgr.add_quantity(vis_timer)
+    basicLogmgr.add_quantity(GCStats())
+
+    # Watches are printed periodically during execution
+    basicLogmgr.add_watches(
+        [
+            "step.max",
+            "t_sim.max",
+            "t_step.max",
+            "t_vis",
+            "t_log",
+            "memory_usage_hwm",
+            "myconst",
+        ]
+    )
+
+    for istep in range(200):
+        basicLogmgr.tick_before()
+
+        dt = uniform(0.01, 0.05)
+        set_dt(basicLogmgr, dt)
+        sleep(dt)
+
+        # Illustrate custom timers
+        if istep % 10 == 0:
+            with vis_timer.start_sub_timer():
+                sleep(0.05)
+
+        if istep == 50:
+            print("FYI: Setting watch interval to 5 seconds.")
+            basicLogmgr.set_watch_interval(5)
+
+        if istep == 150:
+            print("FYI: Setting watch interval back to 1 second.")
+            basicLogmgr.set_watch_interval(1)
+
+        basicLogmgr.tick_after()
+
+
+def test_GCStats(basicLogmgr: LogManager):
+    # will check if the example code breaks from using GCStats
+    # should expand on later
+    # currently ensures that after some time, GC from generation 1
+    # eventually goes into generation 2
+    gcStats = GCStats()
+    basicLogmgr.add_quantity(gcStats)
+
+    outerList = {}
```

Comment:
Could this just be a list instead of a dict?

Reply:
I was able to preallocate the list with the following
```Python
N = 1000
# outerList = {}
outerList = [None] * N
```
Both seem fine to me, whichever is cleaner. Due to the variable name, this seems better.

Reply:
Could you just `append()` the "soon to be last" data, instead of preallocating?

Reply:
Yes, that works.